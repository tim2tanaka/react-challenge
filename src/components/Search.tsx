import { ChangeEvent, useState } from 'react';
import { Data } from '../types';
import { actions } from '../app_state/actions';
import { useAppState, useAppDispatch } from '../app_state/context';

export function Search() {
  const state = useAppState();
  const dispatch = useAppDispatch();

  const [inputText, setInputText] = useState('');

  async function loadData() {
    const url = 'https://restcountries.com/v3.1/name';
    try {
      dispatch({ type: actions.clearData.type, payload: [] });
      dispatch({ type: actions.setIsloading.type });
      dispatch({ type: actions.clearErrorMsg.type });
      const res = await fetch(`${url}/${state.search}`);
      if (res.status !== 200) throw new Error('Country search failed!');
      const data = await res.json();
      const paginationData: Data['data'] = getPaginationData(data);
      dispatch({ type: actions.setData.type, payload: paginationData });
      dispatch({ type: actions.setIsloading.type });
      dispatch({ type: actions.setTotalPages.type, payload: data.length });
    } catch (error: any) {
      if (error?.message)
        dispatch({ type: actions.setErrorMsg.type, payload: error.message });
      dispatch({ type: actions.setIsloading.type });
    }
  }

  function getPaginationData(data: Data['data']) {
    const { page, resultsPerPage } = state.pagination;
    // const minPageResults = 5;
    // const paginationPageResults = data.length < resultsPerPage ? minPageResults : resultsPerPage;
    const resultsPage = page <= 0 ? 1 : page;
    const startIndex = (resultsPage - 1) * resultsPerPage;
    const endIndex = startIndex + resultsPerPage;
    dispatch({ type: actions.setTotalPages.type, payload: data.length });
    return data.slice(startIndex, endIndex);
  }

  function handleSearchInput(event: ChangeEvent<HTMLInputElement>): void {
    setInputText(event.target.value);
    dispatch({ type: actions.setSearch.type, payload: event.target.value });
    if (event.target.value === '') dispatch({ type: actions.clearData.type });
  }

  async function handleBtnClick(btn: string): Promise<void> {
    if (btn === 'clear') {
      dispatch({ type: actions.clearData.type, payload: [] });
      dispatch({ type: actions.clearSearch.type });
      dispatch({ type: actions.clearTotalPages.type });
      setInputText('');
    }
    if (btn === 'search') {
      loadData();
    }
  }

  return (
    <div className="search">
      <label>
        Country search:
        <input
          type="text"
          placeholder={'Search for a country...'}
          onChange={(event) => {
            handleSearchInput(event);
          }}
          className="search-input"
          value={inputText ? inputText : ''}
        />
      </label>
      <button
        onClick={() => {
          handleBtnClick('search');
        }}
        className="button"
      >
        Find
      </button>
      {inputText && (
        <button
          onClick={() => {
            handleBtnClick('clear');
          }}
          className="button"
        >
          Clear
        </button>
      )}
    </div>
  );
}
