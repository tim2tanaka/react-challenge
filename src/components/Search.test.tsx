import React, { ReactNode } from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { Search } from './Search';

interface Props {
  children?: ReactNode;
  onClick: React.MouseEventHandler<HTMLButtonElement>;
}

describe('Search', () => {
  test('render Search', () => {
    const { container } = render(<Search />);
    expect(container).toMatchSnapshot();
  });
  test('renders search button', () => {
    render(<Search />);
    const headerElement = screen.getAllByRole('button');
    expect(headerElement[0]).toHaveTextContent('Find');
  });
  test.skip('search btn clicked, renders error', async () => {
    render(<Search />);
    const button = screen.getByRole('button', { name: 'Find' });
    fireEvent.click(button);
    expect(
      await screen.findByText('Country search failed!')
    ).toBeInTheDocument();
  });
});
