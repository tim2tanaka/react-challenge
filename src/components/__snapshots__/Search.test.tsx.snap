// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search render Search 1`] = `
<div>
  <div
    class="search"
  >
    <label>
      Country search:
      <input
        class="search-input"
        placeholder="Search for a country..."
        type="text"
        value=""
      />
    </label>
    <button
      class="button"
    >
      Find
    </button>
  </div>
</div>
`;
